docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,130 @@
+
+1. List LDAP group UIDs for all OpenShift `Groups`, minus a blacklist
+* List LDAP group UIDs for some whitelist of OpenShift `Groups`
+* List LDAP group UIDs for all LDAP groups, minus a blacklist
+* List LDAP group UIDs for some whitelist of LDAP groups
+
+```go
+// LDAPGroupLister lists the LDAP groups that need to be synced by a job. The LDAPGroupLister needs to
+// be paired with an LDAPGroupDataExtractor that understands the format of the unique identifiers
+// returned to represent the LDAP groups to be synced.
+type LDAPGroupLister interface {
+	ListGroups() (groupUIDs []string, err error)
+}
+```
+
+## Collecting LDAP Members and Metadata
+An `LDAPGroupDataExtractor` gathers information on an LDAP group based on an LDAP group UID. It may cache LDAP responses for responsivity. The approach to implementing this structure will vary with LDAP schema as well as sync job request.
+
+```go
+// LDAPGroupDataExtractor retrieves data about an LDAP group from the LDAP server.
+type LDAPGroupDataExtractor interface {
+	// ExtractMembers returns the list of LDAP first-class user entries that are members of the LDAP
+	// group specified by the groupUID
+	ExtractMembers(groupUID string) (members []*ldap.Entry, err error)
+	// ExtractAnnotations returns the mapping of annotation keys to values that the OpenShift group
+	// representing the LDAP group specified by the groupUID should have
+	ExtractAnnotations(groupUID string) (annotations map[string]string, err error)
+
+	LDAPGroupNameMapper
+}
+
+// LDAPGroupNameMapper maps a groupUID representing an LDAP group to the OpenShift Group name for
+// the resource
+type LDAPGroupNameMapper interface {
+	GroupNameFor(groupUID string) (groupName string, err error)
+}
+```
+
+## Determining OpenShift `Identities` for LDAP Members
+This will be done in an identical way to the mapping for [LDAP password processing](https://github.com/openshift/origin/blob/master/pkg/auth/authenticator/password/ldappassword/ldap.go#L227-243) once the requisite data is collected.
+
+```go
+// LDAPIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity userapi.Identity, err error)
+}
+```
+
+## Determining OpenShift `Users` for OpenShift `Identities`
+A new `UserIdentityMapper` will need to be created: the `DeterministicUserIdentityMapper`, which will assume that the unique identifier retrieved from the LDAP response and integrated into the OpenShift `Identity` will be deterministicically mapped to an OpenShift `User`. This is required as otherwise there is no deterministic mapping from the `Identity` created and any OpenShift `User`.
```

Comment:
> I thought we didn't want to materialize these. I'm fine with using the deterministic mapping without materializing

I'd prefer keep this description of how the mapping is done.  As Steve noted, it doesn't require any materialization.